server/bleep/src/semantic/embedder.rs

```diff
@@ -0,0 +1,271 @@
+            output.push(PointStruct {
+                id: Some(PointId::from(entry.id)),
+                vectors: Some(
+                    tokio::task::block_in_place(|| self.embed(entry.data.as_ref()))?.into(),
+                ),
+                payload: entry.payload,
+            });
+        }
+
+        Ok(output)
+    }
+}
+
+pub struct RemoteEmbedder {
+    url: reqwest::Url,
+    session: reqwest::Client,
+    embedder: LocalEmbedder,
+}
+
+impl RemoteEmbedder {
+    pub fn new(url: reqwest::Url, model_dir: &Path) -> anyhow::Result<Self> {
+        let url = url.join("encode")?;
+        Ok(Self {
+            url,
+            session: reqwest::Client::builder().gzip(true).build()?,
+            embedder: LocalEmbedder::new(model_dir)?,
+        })
+    }
+
+    async fn make_request(&self, request: ServerRequest<'_>) -> anyhow::Result<ServerResponse> {
+        Ok(self
+            .session
+            .post(self.url.clone())
+            .json(&request)
+            .send()
+            .await?
+            .json()
+            .await?)
+    }
+}
+
+#[async_trait]
+impl Embedder for RemoteEmbedder {
+    fn embed(&self, data: &str) -> anyhow::Result<Embedding> {
+        self.embedder.embed(data)
+    }
+
+    fn tokenizer(&self) -> &Tokenizer {
+        self.embedder.tokenizer()
+    }
+
+    async fn batch_embed(&self, log: &EmbedLog, flush: bool) -> anyhow::Result<Vec<PointStruct>> {
+        const MAX_BATCH_SIZE: usize = 128;
+        let mut output = vec![];
+
+        loop {
+            // if we're not currently flushing the log, only process full batches
+            if log.is_empty() || (log.len() < MAX_BATCH_SIZE && !flush) {
+                return Ok(output);
+            }
+
+            let mut batch = vec![];
+
+            // fill this batch with embeddings
+            while let Some(embedding) = log.pop() {
+                batch.push(embedding);
+
+                if batch.len() == MAX_BATCH_SIZE {
+                    break;
+                }
+            }
+
+            let time = Instant::now();
+            let res = self
+                .make_request(ServerRequest {
+                    sequence: batch
+                        .iter()
+                        .map(|embed| embed.data.as_ref())
+                        .collect::<Vec<_>>(),
+                })
+                .await;
+            let elapsed = time.elapsed();
+
+            match res {
+                Ok(res) => {
+                    info!(?elapsed, size = batch.len(), "batch embedding successful");
+                    output.extend(res.data.into_iter().zip(batch).map(|(result, src)| {
+                        PointStruct {
+                            id: Some(PointId::from(src.id)),
+                            vectors: Some(result.embedding.into()),
+                            payload: src.payload,
+                        }
+                    }))
+                }
+                Err(err) => {
+                    error!(
```

Comment:
Can we call local embedder in case remote embedder raises an error. This will make it more reliable.

Reply:
We'll tackle this in a separate PR